### PR TITLE
Add packageIQR repr

### PR DESCRIPTION
**Pull request by janschulz**
These are for output of `data()` or `vignette()`

Closes: https://github.com/IRkernel/repr/issues/7

### CODE REVIEW

File: repr.Rproj [hunk 1]
```diff
@@ -5,14 +5,15 @@ SaveWorkspace: Default
 AlwaysSaveHistory: Default
 
 EnableCodeIndexing: Yes
-UseSpacesForTab: No
+UseSpacesForTab: Yes
```
Comment (flying-sheep) [inline repr.Rproj]: no it doesn’t. all files except for the one you used use tabs. please revert this and fix the file’s indentation

**Comment by flying-sheep** [comment]
woot! cool stuff.

you didn’t use this package’s indentation setting though.

**Comment by janschulz** [comment]
Clean up the tab/space problem... That was because I tested the functions in the notebook and currently sometimes the notebook shows `->`-signs for tabs on copied in code which prevent the code from running... No idea who is to blame here, but it's another part why I really like plain ASCII for code...

**Comment by takluyver** [comment]
Tabs, to be fair, are ASCII. That's bizarre about them in the notebook, though. Do you see the same in a Python notebook?

**Comment by flying-sheep** [comment]
the CodeMirror instance in the notebooks is configured to show tabs.

if you do

```js
%%javascript
CodeMirror.defaults.indentWithTabs = true
```

or

```r
display_javascript('CodeMirror.defaults.indentWithTabs = true')
```

you’ll see it

File: R/repr_packageIQR.r [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+#'
+#' @name repr_*.packageIQR
+#' @export
+repr_text.packageIQR <- function(obj, ...)
+{
+  # this is mostly copied from utils:::print.packageIQR
+  db <- obj$results
+  out <- if (nrow(db) > 0L) {
+	lapply(split(seq_len(nrow(db)), db[, 'Package']),
+		   function(ind) {
```
Comment (flying-sheep) [inline R/repr_packageIQR.r]: indentation is super weird now. can you ensure that regex-searching `^\t*[ ]` has no matches?

File: R/repr_packageIQR.r [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  for (pkg in names(out)) {
+	writeLines(paste0(ifelse(first, '', '\n'), obj$title, ' in package ', sQuote(pkg), ':\n'), outConn)
+	writeLines(formatDL(out[[pkg]][, 'Item'], out[[pkg]][,'Title']), outConn)
+	first <- FALSE
+  }
+#  print(first)
+  if (first) {
+	writeLines(paste('no', tolower(obj$title), 'found'), outConn)
+	if (!is.null(obj$footer))
+	  writeLines(c('', obj$footer), outConn)
```
Comment (flying-sheep) [inline R/repr_packageIQR.r]: also here

File: R/repr_packageIQR.r [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+{
+  db <- as.data.frame(obj$results, stringsAsFactors = FALSE)[c('Package', 'Item', 'Title' )]
+  title = sprintf('<h3>%s</h3>', obj$title)
+  if (nrow(db) == 0) {
+	content = sprintf('<p>No %s found</p>', tolower(obj$title))
+  } else {
+	oldrows = getOption('repr.matrix.max.rows')
+	options(repr.matrix.max.rows = 1000)
+	on.exit({
+	  options(repr.matrix.max.rows = oldrows)
```
Comment (flying-sheep) [inline R/repr_packageIQR.r]: and here

File: R/repr_packageIQR.r [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  }
+#  print(first)
+  if (first) {
+	writeLines(paste('no', tolower(obj$title), 'found'), outConn)
+	if (!is.null(obj$footer))
+	  writeLines(c('', obj$footer), outConn)
+  }
+  else {
+	if (!is.null(obj$footer))
+	  writeLines(c('\n', obj$footer), outConn)
```
Comment (flying-sheep) [inline R/repr_packageIQR.r]: here

File: tests/testthat/test_repr_packageIQR.r [hunk 6]
```diff
@@ -0,0 +1,29 @@
+context('reprs of packageIQR')
+
+options(stringsAsFactors = FALSE)
+
+test_that('repr_html out the right html code', {
+  x = vignette()
```
Comment (flying-sheep) [inline tests/testthat/test_repr_packageIQR.r]: and while you’re at it in this file. indentation settings are per project in Rstudio so it’ll cause pain if a file is indented differently

File: R/repr_packageIQR.r [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  }
+  close(outConn)
+  #print(output)
+  #print('Damn')
```
Comment (flying-sheep) [inline R/repr_packageIQR.r]: pls no commented-out code

File: R/repr_packageIQR.r [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  outConn = textConnection('output', 'w', local = TRUE)
+  
+  first <- TRUE
+  
+  for (pkg in names(out)) {
+	writeLines(paste0(ifelse(first, '', '\n'), obj$title, ' in package ', sQuote(pkg), ':\n'), outConn)
+	writeLines(formatDL(out[[pkg]][, 'Item'], out[[pkg]][,'Title']), outConn)
+	first <- FALSE
+  }
+#  print(first)
```
Comment (flying-sheep) [inline R/repr_packageIQR.r]: pls no commented-out code

File: R/repr_packageIQR.r [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	on.exit({
+	  options(repr.matrix.max.rows = oldrows)
+	})
+	content <- repr_html(db)
+  }
+  footer <- ''
+  if (!is.null(obj$footer)) {
+	footer <- sprintf('<p>%s</p>', obj$footer)
+  }
+  paste(title, '\n', content, '\n', footer, collapse = '\n')
```
Comment (flying-sheep) [inline R/repr_packageIQR.r]: as said: i think collapse only works as intended for one object.

i think it merges everything with `sep` and *then* collapses. are you sure this works?

File: R/repr_packageIQR.r (hunk 2 above)
Comment (janschulz) [inline R/repr_packageIQR.r]: I reformatted both new files in RStudio... As far as I can see, everything is now tabs.

Comment (janschulz) [inline R/repr_packageIQR.r]: I think the problem was, that RSudio has tab = 2 spaces, but my replacer was notepad++ which uses 4 spaces per tab...

File: R/repr_packageIQR.r (hunk 8 above)
Comment (janschulz) [inline R/repr_packageIQR.r]: done

File: R/repr_packageIQR.r (hunk 9 above)
Comment (janschulz) [inline R/repr_packageIQR.r]: fixed... It doesn't need collapse at all I think: all three objects are 1d

File: R/repr_packageIQR.r [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+#'
+#' @name repr_*.packageIQR
+#' @export
+repr_text.packageIQR <- function(obj, ...)
+{
+	# this is mostly copied from utils:::print.packageIQR
+	db <- obj$results
+	out <- if (nrow(db) > 0L) {
+		lapply(split(seq_len(nrow(db)), db[, 'Package']),
+					 function(ind) {
```
Comment (flying-sheep) [inline R/repr_packageIQR.r]: OMG :flushed: 

…i guess i’ll just do the formatting afterwards, OK?

File: R/repr_packageIQR.r [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+	title = sprintf('<h3>%s</h3>', obj$title)
+	if (nrow(db) == 0) {
+		content = sprintf('<p>No %s found</p>', tolower(obj$title))
+	} else {
+		oldrows = getOption('repr.matrix.max.rows')
+		options(repr.matrix.max.rows = 1000)
+		on.exit({
+			options(repr.matrix.max.rows = oldrows)
+		})
+		content <- repr_html(db)
```
Comment (flying-sheep) [inline R/repr_packageIQR.r]: nice :smile:

**Comment by flying-sheep** [comment]
i made the matrix repr functions accept rows/cols kwargs in 9ff2891baeb09b4b04167a8cafd5456d3f6e3688 and then improved the code style/simplicity in 901d41a1875aaf68114087b47e24a6723eb28829

**Comment by janschulz** [comment]
Thanks...

**Comment by flying-sheep** [comment]
i hope you don’t take this as criticism!

your idea to simply use `repr_html.data.frame` was great and i didn’t change much in the code you wrote:

as you said, the first was mainly copied from utils, so i did there what you did (coercing to data.frame) and removed the textConnection in favor of appending to the vector.

in the second one i leveraged the new `rows` argument and the fact that `is.null(sprintf('%s', NULL))`.

**Comment by janschulz** [comment]
No, all is good...

I actually thought about using string concat instead of the text connection but thought that using the text connection as string buffer was faster... But you vectorized so that isn't an issue....

The only thing I have is that it would be easier to simple use rstudios code style and try to lint that (if there is a linter which does that...)...

**Comment by flying-sheep** [comment]
it’s probably still a bit slower, but who cares about the speed of like 4 list concatenations.

**Comment by janschulz** [comment]
My non vectorized version would have been a lot more :-)

**Pull request merged**